README.md
# wf

[![latest](https://img.shields.io/github/v/release/bow/wf?labelColor=4d4d4d&color=007c5b&style=flat)](https://github.com/bow/wf/releases)
[![doc](https://img.shields.io/badge/godoc-reference-007c5b?style=flat)](https://godoc.org/github.com/bow/wf/wait)
[![ci](https://img.shields.io/travis/bow/wf?labelColor=4d4d4d&color=007c5b&style=flat)](https://travis-ci.org/bow/wf)
[![cov](https://img.shields.io/codeclimate/coverage/bow/wf?labelColor=4d4d4d&color=007c5b&style=flat)](https://codeclimate.com/github/bow/wf)
[![qual](https://img.shields.io/codeclimate/maintainability/bow/wf?labelColor=4d4d4d&color=007c5b&style=flat)](https://codeclimate.com/github/bow/wf)

`wf` waits until TCP server(s) are ready to accept connections.


## Why?

The main use case for `wf` is to make containerized applications that depend
on external services more robust by waiting for those actual services to be
ready, prior to application start. `wf` can wait on multiple TCP servers and
is provided as a single, static binary for linux/amd64, so it can be added
into the container and used as-is.


## Usage

`wf` is provided as a command line application.

    $ wf --help

    Wait until TCP server(s) are ready to accept connections

    Usage:
      wf [FLAGS] ADDRESS...

    Flags:
      -t, --timeout duration     set wait timeout (default 5s)
      -f, --poll-freq duration   set connection poll frequency (default 500ms)
      -q, --quiet                suppress waiting messages
      -h, --help                 help for wf
          --version              version for wf

The functionalities themselves are provided as a Go library in the
[wait](https://godoc.org/github.com/bow/wait-for/wait) package. Refer to the
relevant GoDoc documentation for a complete documentation.


## Development

`wf` was developed using Go go1.13 on linux/amd64. Other versions and/or
platforms may work but have not been tested.

    # Clone the repository.
    $ git clone https://github.com/bow/wf

    # Run the tests.
    $ make test

    # Build the wf binary.
    $ go build


## Credits

`wf` was inspired by the [wait-for-it.sh](https://github.com/vishnubob/wait-for-it) script by @vishnubob.


CHANGELOG.md
# Changelog


## Version 0.0.0

`release date: 2020-02-01`

* First release.


cmd/cmd_test.go
package cmd

import (
	"testing"
	"time"
)

func TestRun(t *testing.T) {
	t.Parallel()

	retCode := run([]string{"golang.org:443"}, 5*time.Second, 500*time.Millisecond, false)

	if retCode != 0 {
		t.Errorf("test failed - want exit code: %d, got: %d", 0, retCode)
	}
}


wait/tcp_test.go
package wait

import (
	"context"
	"fmt"
	"net"
	"strconv"
	"sync"
	"testing"
	"time"
)

func TestMessageTarget(t *testing.T) {
	t.Parallel()

	var tests = []struct {
		name string
		in   Message
		want string
	}{
		{
			"with TCPSpec",
			newTCPMessageReady(
				&TCPSpec{Host: "localhost", Port: "7000", PollFreq: 1 * time.Second},
				time.Now(),
			),
			"tcp://localhost:7000",
		},
		{
			"no TCPSpec",
			newTCPMessageFailed(nil, time.Now(), fmt.Errorf("stub")),
			"<none>",
		},
	}

	for i, test := range tests {
		i := i
		test := test

		t.Run(test.name, func(t *testing.T) {
			t.Parallel()

			name := test.name
			want := test.want
			got := test.in.Target()

			if want != got {
				t.Errorf("test[%d] %q failed - want: %q, got: %q", i, name, want, got)
			}
		})
	}
}

func TestParseTCPSpec(t *testing.T) {
	t.Parallel()

	var commonPollFreq = 1 * time.Second
	var tests = []struct {
		name     string
		in       string
		wantSpec *TCPSpec
		wantErr  error
	}{
		{
			"no protocol, no port",
			"localhost",
			nil,
			fmt.Errorf("neither port nor protocol is given"),
		},
		{
			"unknown protocol, no port",
			"foo://localhost",
			nil,
			fmt.Errorf("port not given and protocol is unknown: \"foo\""),
		},
		{
			"no protocol, port, no poll freq",
			"localhost:5000",
			&TCPSpec{
				Host:     "localhost",
				Port:     "5000",
				PollFreq: commonPollFreq,
			},
			nil,
		},
		{
			"no protocol, port, poll freq",
			"localhost:5000#3s",
			&TCPSpec{
				Host:     "localhost",
				Port:     "5000",
				PollFreq: 3 * time.Second,
			},
			nil,
		},
		{
			"http, no port, no poll freq",
			"http://localhost",
			&TCPSpec{
				Host:     "localhost",
				Port:     "80",
				PollFreq: commonPollFreq,
			},
			nil,
		},
		{
			"http, no port, poll freq",
			"http://localhost#500ms",
			&TCPSpec{
				Host:     "localhost",
				Port:     "80",
				PollFreq: 500 * time.Millisecond,
			},
			nil,
		},
		{
			"http, port, no poll freq",
			"http://localhost:3000",
			&TCPSpec{
				Host:     "localhost",
				Port:     "3000",
				PollFreq: commonPollFreq,
			},
			nil,
		},
		{
			"http, port, poll freq",
			"http://localhost:3000#2s",
			&TCPSpec{
				Host:     "localhost",
				Port:     "3000",
				PollFreq: 2 * time.Second,
			},
			nil,
		},
	}

	for i, test := range tests {
		i := i
		test := test

		t.Run(test.name, func(t *testing.T) {
			t.Parallel()

			name := test.name
			wantSpec := test.wantSpec
			wantErr := test.wantErr
			gotSpec, gotErr := ParseTCPSpec(test.in, commonPollFreq)

			if wantErr != nil && gotErr.Error() != wantErr.Error() {
				t.Errorf("test[%d] %q failed - want err: %q, got: %q", i, name, wantErr, gotErr)
			}

			if wantErr == nil && *wantSpec != *gotSpec {
				t.Errorf(
					"test[%d] %q failed - want spec: %+v, got: %+v",
					i,
					name,
					*wantSpec,
					*gotSpec,
				)
			}
		})
	}
}

func ExampleParseTCPSpec() {
	spec, _ := ParseTCPSpec("golang.org:80", 1*time.Second)
	fmt.Println("host:", spec.Host)
	fmt.Println("port:", spec.Port)
	fmt.Println("poll freq:", spec.PollFreq)
	// Output:
	// host: golang.org
	// port: 80
	// poll freq: 1s
}

func ExampleParseTCPSpec_proto() {
	spec, _ := ParseTCPSpec("https://golang.org", 1*time.Second)
	fmt.Println("host:", spec.Host)
	fmt.Println("port:", spec.Port)
	fmt.Println("poll freq:", spec.PollFreq)
	// Output:
	// host: golang.org
	// port: 443
	// poll freq: 1s
}

func ExampleParseTCPSpec_freq() {
	spec, _ := ParseTCPSpec("amqps://127.0.0.1#500ms", 1*time.Second)
	fmt.Println("host:", spec.Host)
	fmt.Println("port:", spec.Port)
	fmt.Println("poll freq:", spec.PollFreq)
	// Output:
	// host: 127.0.0.1
	// port: 5671
	// poll freq: 500ms
}

func TestParseTCPSpecs(t *testing.T) {
	t.Parallel()

	var commonPollFreq = 1 * time.Second
	var tests = []struct {
		name      string
		in        []string
		wantSpecs []*TCPSpec
		wantErr   error
	}{
		{
			"all ok",
			[]string{
				"127.0.0.1:3000",
				"https://golang.org",
				"localhost:1234#200ms",
			},
			[]*TCPSpec{
				{"127.0.0.1", "3000", 1 * time.Second},
				{"golang.org", "443", 1 * time.Second},
				{"localhost", "1234", 200 * time.Millisecond},
			},
			nil,
		},
		{
			"some err",
			[]string{
				"127.0.0.1:3000",
				"localhost",
				"localhost:1234#200ms",
			},
			[]*TCPSpec{},
			fmt.Errorf("address 1: neither port nor protocol is given"),
		},
	}

	for i, test := range tests {
		i := i
		test := test
		t.Run(test.name, func(t *testing.T) {
			t.Parallel()

			name := test.name
			wantSpecs := test.wantSpecs
			wantErr := test.wantErr

			gotSpecs, gotErr := ParseTCPSpecs(test.in, commonPollFreq)

			if wantErr != nil && gotErr.Error() != wantErr.Error() {
				t.Errorf("test[%d] %q failed - want error: %q, got: %q", i, name, wantErr, gotErr)
			}

			if len(wantSpecs) != len(gotSpecs) {
				t.Fatalf(
					"test[%d] %q failed - want: %d specs, got: %d",
					i,
					name,
					len(wantSpecs),
					len(gotSpecs),
				)
			}
			for j, wantSpec := range wantSpecs {
				gotSpec := gotSpecs[j]
				if wantErr == nil && *wantSpec != *gotSpec {
					t.Errorf(
						"test[%d][%d] %q failed - got spec: %+v, want: %+v",
						i,
						j,
						name,
						*gotSpec,
						*wantSpec,
					)
				}
			}
		})
	}
}

// tcpServerHost is the hostname for the test TCP server.
const tcpServerHost = "127.0.0.1"

// getLocalTCPPort returns a TCP port for testing by asking the kernel for a free port.
func getLocalTCPPort() string {
	addr, err := net.ResolveTCPAddr("tcp", net.JoinHostPort(tcpServerHost, "0"))
	if err != nil {
		panic(err)
	}

	listener, err := net.ListenTCP("tcp", addr)
	if err != nil {
		panic(err)
	}
	defer listener.Close()

	return strconv.Itoa(listener.Addr().(*net.TCPAddr).Port)
}

// tcpServer is a wrapper struct for launching test TCP servers.
type tcpServer struct {
	host, port string
	// readyDelay is the duration to wait before the server is running.
	readyDelay time.Duration
	t          *testing.T
}

// addr returns the tcpServer address.
func (srv *tcpServer) addr() string {
	return net.JoinHostPort(srv.host, srv.port)
}

// start starts the test TCP server. It returns a context.Context value based on the input context,
// along with a cancellation function for stopping the server and ensuring proper cleanup.
func (srv *tcpServer) start(ctx context.Context) (context.Context, context.CancelFunc) {
	ictx, icancel := context.WithCancel(ctx)

	go func(addr string, delay time.Duration, gctx context.Context, t *testing.T) {
		select {
		// Handle case when the goroutine needs to be killed prior to server start.
		case <-gctx.Done():
			return
		// Expected flow: wait for `delay` before starting the server.
		case <-time.After(delay):
		}

		listener, err := net.Listen("tcp", addr)
		if err != nil {
			t.Logf("failed starting test TCP server %q: %s", addr, err)
			return
		}
		defer listener.Close()

		for {
			conn, err := listener.Accept()
			if err != nil {
				t.Logf("failed accepting TCP connection %q: %s", addr, err)
				return
			}
			defer conn.Close()
			select {
			case <-gctx.Done():
				return
			default:
			}
		}
	}(srv.addr(), srv.readyDelay, ictx, srv.t)

	return ictx, func() {
		var addr = srv.addr()
		icancel()
		// Dial to the server so that listener.Accept progresses and the ctx.Done() case is
		// selected.
		conn, err := net.Dial("tcp", addr)
		if err != nil {
			return
		}
		conn.Close()
	}
}

// tcpServerGroup is a helper container for starting multiple TCP servers.
type tcpServerGroup struct {
	servers []*tcpServer
	t       *testing.T
}

// start starts all the TCP servers in the group, ensuring they do so at the same time. It returns a
// context.Context based on the input context, along with a cancellation function for stopping all
// the servers and ensuring proper cleanup.
func (grp *tcpServerGroup) start(ctx context.Context) (context.Context, context.CancelFunc) {
	var (
		wgStart, wgEnd sync.WaitGroup
		ictx, icancel  = context.WithCancel(ctx)
	)

	// Track start and end jobs.
	wgStart.Add(1)
	wgEnd.Add(1)

	for _, srv := range grp.servers {
		go func(srv *tcpServer, ictx context.Context, wgStart, wgEnd *sync.WaitGroup) {
			wgStart.Wait()
			_, cancel := srv.start(ictx)
			// Wait until outer scope calls wgEnd.Done.
			wgEnd.Wait()
			cancel()
		}(srv, ictx, &wgStart, &wgEnd)
	}
	// Start all servers at the same time.
	wgStart.Done()

	return ictx, func() {
		icancel()
		// Release wgEnd.Wait() block in all launched goroutines.
		wgEnd.Done()
	}
}

// messageBox is a test helper container for messages emitted by the wait operations.
type messageBox struct {
	msgs []Message
}

// newMessageBox creates a messageBox by draining all the messages from the given channel.
func newMessageBox(ch <-chan *TCPMessage) *messageBox {
	msgs := make([]Message, 0)
	for msg := range ch {
		msgs = append(msgs, msg)
	}
	return &messageBox{msgs: msgs}
}

// count returns the number of messages in the box.
func (mb *messageBox) count() int {
	return len(mb.msgs)
}

// filterByTCPAddr returns a new message box containing only TCPMessages with the given address.
func (mb *messageBox) filterByTCPAddr(addr string) *messageBox {
	filtered := make([]Message, 0)
	for _, msg := range mb.msgs {
		if tcpMsg, isTCPMessage := msg.(*TCPMessage); isTCPMessage && tcpMsg.Addr() == addr {
			filtered = append(filtered, tcpMsg)
		}
	}
	return &messageBox{msgs: filtered}
}

func TestOneTCPReady(t *testing.T) {
	t.Parallel()

	var (
		waitTimeout = 3 * time.Second
		server      = &tcpServer{
			host:       tcpServerHost,
			port:       getLocalTCPPort(),
			readyDelay: 1 * time.Second,
			t:          t,
		}
		spec = &TCPSpec{Host: server.host, Port: server.port, PollFreq: 500 * time.Millisecond}
	)

	_, cancel := server.start(context.Background())
	defer cancel()

	msgs := OneTCP(spec, waitTimeout)

	// There must be 2 messages in total.
	mb := newMessageBox(msgs)
	if msgCount := mb.count(); msgCount != 2 {
		t.Fatalf("test failed - want %d messages, got %d", 2, msgCount)
	}

	// The last message's ElapsedTime must be at least equal to waitTimeout.
	if elTime := mb.msgs[mb.count()-1].ElapsedTime(); elTime >= waitTimeout {
		t.Errorf("test failed - elapsed time %s exceeded timeout limit of %s", elTime, waitTimeout)
	}

	// The messages from waiting for the server must be as expected.
	if status := mb.msgs[0].Status(); status != Start {
		t.Errorf("test msgs[0].Status() failed - want: %s, got %s", Start, status)
	}
	if status := mb.msgs[1].Status(); status != Ready {
		t.Errorf("test msgs[1].Status() failed - want: %s, got %s", Ready, status)
	}
}

func TestAllTCPReady(t *testing.T) {
	t.Parallel()

	var (
		waitTimeout = 5 * time.Second
		servers     = []*tcpServer{
			{tcpServerHost, getLocalTCPPort(), 0 * time.Second, t},
			{tcpServerHost, getLocalTCPPort(), 3 * time.Second, t},
		}
		group = tcpServerGroup{servers: servers, t: t}
	)

	_, cancel := group.start(context.Background())
	defer cancel()

	msgs := AllTCP(
		[]*TCPSpec{
			{servers[0].host, servers[0].port, 500 * time.Millisecond},
			{servers[1].host, servers[1].port, 500 * time.Millisecond},
		},
		waitTimeout,
	)

	// There must be 4 messages in total.
	mb := newMessageBox(msgs)
	if msgCount := mb.count(); msgCount != 4 {
		t.Fatalf("test failed - want %d messages, got %d", 4, msgCount)
	}

	// The last message's ElapsedTime must be less than waitTimeout.
	if elTime := mb.msgs[mb.count()-1].ElapsedTime(); elTime >= waitTimeout {
		t.Errorf("test failed - elapsed time %s exceeded timeout limit of %s", elTime, waitTimeout)
	}

	// The messages from waiting for the first server must be as expected.
	addr1 := servers[0].addr()
	mb1 := mb.filterByTCPAddr(addr1)
	if msgCount := mb1.count(); msgCount != 2 {
		t.Fatalf("test[%s] failed - want %d messages, got %d", addr1, 2, msgCount)
	}
	if status := mb1.msgs[0].Status(); status != Start {
		t.Errorf("test[%s] msgs[0].Status() failed - want: %s, got %s", addr1, Start, status)
	}
	if status := mb1.msgs[1].Status(); status != Ready {
		t.Errorf("test[%s] msgs[1].Status() failed - want: %s, got %s", addr1, Ready, status)
	}

	// The messages from waiting for the second server must be as expected.
	addr2 := servers[1].addr()
	mb2 := mb.filterByTCPAddr(addr2)
	if msgCount := mb2.count(); msgCount != 2 {
		t.Fatalf("test[%s] failed - want %d messages, got %d", addr2, 2, msgCount)
	}
	if status := mb2.msgs[0].Status(); status != Start {
		t.Errorf("test[%s] msgs[0].Status() failed - want: %s, got %s", addr2, Start, status)
	}
	if status := mb2.msgs[1].Status(); status != Ready {
		t.Errorf("test[%s] msgs[1].Status() failed - want: %s, got %s", addr2, Ready, status)
	}
}

func TestAllTCPTimeout(t *testing.T) {
	t.Parallel()

	var (
		waitTimeout = 5 * time.Second
		servers     = []*tcpServer{
			{tcpServerHost, getLocalTCPPort(), 10 * time.Second, t},
			{tcpServerHost, getLocalTCPPort(), 1 * time.Second, t},
		}
		group = tcpServerGroup{servers: servers, t: t}
	)

	_, cancel := group.start(context.Background())
	defer cancel()

	msgs := AllTCP(
		[]*TCPSpec{
			{servers[0].host, servers[0].port, 500 * time.Millisecond},
			{servers[1].host, servers[1].port, 500 * time.Millisecond},
		},
		waitTimeout,
	)

	// There must be 4 messages in total.
	mb := newMessageBox(msgs)
	if msgCount := mb.count(); msgCount != 4 {
		t.Fatalf("test failed - want %d messages, got %d", 4, msgCount)
	}

	// The last message's ElapsedTime must be at least equal to waitTimeout.
	if elTime := mb.msgs[mb.count()-1].ElapsedTime(); elTime < waitTimeout {
		t.Errorf("test failed - elapsed time %s is less than timeout limit of %s", elTime, waitTimeout)
	}
	// The last one must be a timeout failure.
	if status := mb.msgs[mb.count()-1].Status(); status != Failed {
		t.Errorf("test failed msgs[-1].Status() failed - want: %s, got: %s", Failed, status)
	}

	// The messages from waiting for the first server must be as expected.
	addr1 := servers[0].addr()
	mb1 := mb.filterByTCPAddr(addr1)
	if msgCount := mb1.count(); msgCount != 1 {
		t.Fatalf("test[%s] failed - want: %d messages, got: %d", addr1, 1, msgCount)
	}
	if status := mb1.msgs[0].Status(); status != Start {
		t.Errorf("test[%s] msgs[0].Status() failed - want: %s, got: %s", addr1, Start, status)
	}

	// The messages from waiting for the second server must be as expected.
	addr2 := servers[1].addr()
	mb2 := mb.filterByTCPAddr(addr2)
	if msgCount := mb2.count(); msgCount != 2 {
		t.Fatalf("test[%s] failed - want: %d messages, got: %d", addr2, 2, msgCount)
	}
	if status := mb2.msgs[0].Status(); status != Start {
		t.Errorf("test[%s] msgs[0].Status() failed - want: %s, got %s", addr2, Start, status)
	}
	if status := mb2.msgs[1].Status(); status != Ready {
		t.Errorf("test[%s] msgs[1].Status() failed - want: %s, got %s", addr2, Ready, status)
	}
}


cmd/utils_test.go
package cmd

import (
	"testing"
	"time"
)

func TestFmtElapsedTime(t *testing.T) {
	t.Parallel()

	var tests = []struct {
		in   time.Duration
		want string
	}{
		{0 * time.Nanosecond, "0s"},
		{45 * time.Nanosecond, "45ns"},
		{24313 * time.Nanosecond, "24.31µs"},
		{759825 * time.Nanosecond, "759.83µs"},
		{999995 * time.Nanosecond, "1ms"},
		{999994 * time.Nanosecond, "999.99µs"},
		{32423 * time.Microsecond, "32.42ms"},
		{301451654 * time.Microsecond, "5m1.45s"},
		{287336 * time.Millisecond, "4m47.34s"},
		{125432 * time.Millisecond, "2m5.43s"},
		{301 * time.Second, "5m1s"},
	}

	for i, test := range tests {
		i := i
		test := test
		name := test.in.String()

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			want := test.want
			got := fmtElapsedTime(test.in)

			if want != got {
				t.Errorf("test[%d] %q failed - want: %q, got: %q", i, name, want, got)
			}
		})
	}
}


cmd/cmd.go
// Package cmd provides the command line interface for the
// https://godoc.org/github.com/bow/wf/wait package.
package cmd

import (
	"fmt"
	"os"
	"time"

	"github.com/spf13/cobra"

	"github.com/bow/wf/wait"
)

const (
	name = "wf"
	desc = "Wait until TCP server(s) are ready to accept connections"
)

var (
	// These are meant to be overidden at built time using ldflags -X.
	buildTime = "?"
	version   = "dev"
)

// Execute peforms the actual CLI argument parsing and launches the wait operation.
func Execute() error {
	var (
		waitTimeout     time.Duration
		defaultPollFreq time.Duration
		isQuiet         bool
	)

	cmd := &cobra.Command{
		Use:                   name + " [FLAGS] ADDRESS...",
		Short:                 desc,
		Version:               version + " (build time " + buildTime + ")",
		DisableFlagsInUseLine: true,

		Args: func(cmd *cobra.Command, args []string) error {
			if len(args) < 1 {
				return fmt.Errorf("at least one address must be specified")
			}
			return nil
		},

		Run: func(cmd *cobra.Command, args []string) {
			var rawAddrs []string
			if dashIdx := cmd.ArgsLenAtDash(); dashIdx == -1 {
				rawAddrs = args
			} else {
				rawAddrs = args[:dashIdx]
			}
			exitCode := run(rawAddrs, waitTimeout, defaultPollFreq, isQuiet)
			if exitCode != 0 {
				os.Exit(exitCode)
			}
		},
	}

	flagSet := cmd.Flags()
	flagSet.SortFlags = false
	flagSet.DurationVarP(&waitTimeout, "timeout", "t", 5*time.Second, "set wait timeout")
	flagSet.DurationVarP(
		&defaultPollFreq,
		"poll-freq",
		"f",
		500*time.Millisecond,
		"set connection poll frequency",
	)
	flagSet.BoolVarP(&isQuiet, "quiet", "q", false, "suppress waiting messages")

	return cmd.Execute()
}

// run calls the actual function for waiting.
func run(
	rawAddrs []string,
	waitTimeout, defaultPollFreq time.Duration,
	isQuiet bool,
) int {

	specs, err := wait.ParseTCPSpecs(rawAddrs, defaultPollFreq)
	if err != nil {
		fmt.Printf("%7s: %s\n", "ERROR", err)
		return 1
	}

	var (
		msg       wait.Message
		showMsg   = func(wait.Message) {}
		showFinal = func(wait.Message) {}
	)
	if !isQuiet {
		showMsg = func(msg wait.Message) {
			var disp string

			switch msg.Status() {
			case wait.Start:
				disp = fmt.Sprintf("%7s: %s for %s", "waiting", msg.Target(), waitTimeout)
			case wait.Ready:
				disp = fmt.Sprintf(
					"%7s: %s in %s",
					wait.Ready,
					msg.Target(),
					fmtElapsedTime(msg.ElapsedTime()),
				)
			case wait.Failed:
				disp = fmt.Sprintf("%7s: %s", wait.Failed, msg.Err())
			}

			fmt.Println(disp)
		}
		showFinal = func(msg wait.Message) {
			fmt.Printf("%7s: all ready in %s\n", "OK", fmtElapsedTime(msg.ElapsedTime()))
		}
	}

	for msg = range wait.AllTCP(specs, waitTimeout) {
		showMsg(msg)
		if err := msg.Err(); err != nil {
			return 1
		}
	}
	showFinal(msg)

	return 0
}


wait/tcp.go
package wait

import (
	"context"
	"fmt"
	"net"
	"regexp"
	"strings"
	"time"
)

var (
	// addrPattern is used for parsing input TCP addresses and extracting the relevant parts.
	addrPattern = regexp.MustCompile(
		"^(?P<schema>(?P<proto>[A-Za-z]+)://)?(?P<host>[^#]+)(#(?P<freq>.+))?",
	)
	// protoPort is a mapping between popular TCP-backed protocol names to their default port
	// numbers.
	protoPort = map[string]string{
		"amqp":       "5672",
		"amqps":      "5671",
		"http":       "80",
		"https":      "443",
		"imap":       "143",
		"mysql":      "3306",
		"ldap":       "389",
		"ldaps":      "636",
		"postgresql": "5432",
		"smtp":       "25",
	}
)

// TCPSpec represents the input specification of a single TCP wait operation.
type TCPSpec struct {
	// Host is the hostname or IP address being waited.
	Host string
	// Port is the port number for the connection.
	Port string
	// PollFreq is how often a connection is attempted.
	PollFreq time.Duration
}

// Addr returns the host and port of the TCP specifications, joined by ':'.
func (spec *TCPSpec) Addr() string {
	return net.JoinHostPort(spec.Host, spec.Port)
}

// Message is the interface for messages sent by the wait operations.
type Message interface {
	// Status returns the status of the message.
	Status() Status
	// Target returns the entity being waited.
	Target() string
	// Err returns an error, if the message contains any.
	Err() error
	// ElapsedTime returns the duration of the wait operation at the time of message creation.
	ElapsedTime() time.Duration
}

// TCPMessage is a container for wait operations on TCP servers.
type TCPMessage struct {
	// spec is the wait operation specifications.
	spec *TCPSpec
	// status is the wait operation status.
	status Status
	// startTime is when the wait operation starts.
	startTime time.Time
	// emitTime is when the message is created and emitted. The current implementation creates and
	// emits at the same time.
	emitTime time.Time
	// err is any operation that may have occurred.
	err error
}

// newTCPMessageStart creates a new TCPMessage with status Start and no errors.
func newTCPMessageStart(spec *TCPSpec, startTime time.Time) *TCPMessage {
	return &TCPMessage{
		spec:      spec,
		status:    Start,
		startTime: startTime,
		emitTime:  time.Now(),
		err:       nil,
	}
}

// newTCPMessageReady creates a new TCPMessage with status Ready and no errors.
func newTCPMessageReady(spec *TCPSpec, startTime time.Time) *TCPMessage {
	return &TCPMessage{
		spec:      spec,
		status:    Ready,
		startTime: startTime,
		emitTime:  time.Now(),
		err:       nil,
	}
}

// newTCPMessage failed creates a new TCPMessage with status Failed and the given error.
func newTCPMessageFailed(spec *TCPSpec, startTime time.Time, err error) *TCPMessage {
	return &TCPMessage{
		spec:      spec,
		status:    Failed,
		startTime: startTime,
		emitTime:  time.Now(),
		err:       err,
	}
}

// Status returns the status of the message.
func (msg *TCPMessage) Status() Status {
	return msg.status
}

// Target returns the target of the wait operation, which is `tcp://` prepended to Addr. If the
// specifications is nil, this returns `<none>`.
func (msg *TCPMessage) Target() string {
	if msg.spec == nil {
		return "<none>"
	}
	return "tcp://" + msg.Addr()
}

// Addr returns the address being waited. If the specifications is nil, this returns `<none>`.
func (msg *TCPMessage) Addr() string {
	if msg.spec == nil {
		return "<none>"
	}
	return msg.spec.Addr()
}

// ElapsedTime is the duration between waiting operation start and status emission.
func (msg *TCPMessage) ElapsedTime() time.Duration {
	return msg.emitTime.Sub(msg.startTime)
}

// Err returns the error contained in the message, if any.
func (msg *TCPMessage) Err() error {
	return msg.err
}

// ctxKey is the key type for wait contexts.
type ctxKey int

// startTimeCtxKey is the key for retrieving wait operation start time from contexts.
const startTimeCtxKey ctxKey = 0

// newContext creates a new context containing current time along with a cancellation function,
// based on the background context.
func newContext() (context.Context, context.CancelFunc) {
	ctx, cancel := context.WithCancel(context.Background())
	return context.WithValue(ctx, startTimeCtxKey, time.Now()), cancel
}

// startTimeFromContext extracts the wait operation start time from the given context. If the
// expected value does not exist or it does not typecheck, the current time is returned.
func startTimeFromContext(ctx context.Context) time.Time {
	startTime, ok := ctx.Value(startTimeCtxKey).(time.Time)
	if !ok {
		return time.Now()
	}
	return startTime
}

// ParseTCPSpec parses the given address into a TCPSpec and then returns a pointer to it. The
// address can be given in several forms: `<host>:<port>`, `<protocol>://<host>`, or
// `<protocol>://<host>:<port>`. For the second form, if the protocol is known, the port will be
// inferred from it (e.g. port 80 for HTTP and 443 for HTTPS). For the last form, the `<protocol>`
// is ignored.  This function also takes a `defaultPollFreq` argument, which it will use as the poll
// frequency of the TCPSpec if the raw address does not specify a poll frequency value.  The poll
// frequency value in the raw address is the string value of time.Duration, appended to the address
// after a `#` sign.
func ParseTCPSpec(rawAddr string, defaultPollFreq time.Duration) (*TCPSpec, error) {
	var (
		proto             string
		rawHost           string
		hasPort, hasProto bool
		matches           = addrPattern.FindStringSubmatch(rawAddr)
		subexpNames       = addrPattern.SubexpNames()
		groups            = make(map[string]string)
	)

	for i, value := range matches {
		groups[subexpNames[i]] = value
	}

	rawHost = groups["host"]
	hasPort = strings.ContainsRune(rawHost, ':')

	if hasPort {
		host, port, err := net.SplitHostPort(rawHost)
		if err != nil {
			return nil, err
		}
		groups["host"] = host
		groups["port"] = port
	} else if proto, hasProto = groups["proto"]; hasProto {
		if port, knownProto := protoPort[strings.ToLower(proto)]; knownProto {
			groups["host"] = rawHost
			groups["port"] = port
		} else {
			if proto == "" {
				return nil, fmt.Errorf("neither port nor protocol is given")
			}
			return nil, fmt.Errorf("port not given and protocol is unknown: %q", proto)
		}
	}

	if rawFreq, hasFreq := groups["freq"]; hasFreq && rawFreq != "" {
		freq, err := time.ParseDuration(rawFreq)
		if err != nil {
			return nil, err
		}
		defaultPollFreq = freq
	}

	return &TCPSpec{
		Host:     groups["host"],
		Port:     groups["port"],
		PollFreq: defaultPollFreq,
	}, nil
}

// ParseTCPSpecs parses multiple addresses into separate TCPSpecs, returned as a slice of pointers.
// It has the same semantics as `ParseTCPSpec`, only it works with multiple addresses instead of
// one.
func ParseTCPSpecs(rawAddrs []string, defaultPollFreq time.Duration) ([]*TCPSpec, error) {
	specs := make([]*TCPSpec, len(rawAddrs))

	for i, rawAddr := range rawAddrs {
		spec, err := ParseTCPSpec(rawAddr, defaultPollFreq)
		if err != nil {
			return []*TCPSpec{}, fmt.Errorf("address %d: %s", i, err)
		}
		specs[i] = spec
	}

	return specs, nil
}

// singleTCP is a helper function for checking TCP server status that accepts a cancellable parent
// context, along with specifications of which server to poll.
func singleTCP(ctx context.Context, spec *TCPSpec) <-chan *TCPMessage {
	startTime := startTimeFromContext(ctx)
	out := make(chan *TCPMessage, 2)

	checkConn := func() *TCPMessage {
		_, err := net.DialTimeout("tcp", spec.Addr(), spec.PollFreq)

		if err == nil {
			return newTCPMessageReady(spec, startTime)
		}
		if shouldWait(err) {
			return nil
		}
		return newTCPMessageFailed(spec, startTime, err)
	}

	go func() {
		pollTicker := time.NewTicker(spec.PollFreq)
		defer pollTicker.Stop()

		defer close(out)

		out <- newTCPMessageStart(spec, startTime)

		// So that we start polling immediately, without waiting for the first tick.
		// There is no way to do this via the current ticker API.
		// See: https://github.com/golang/go/issues/17601
		if msg := checkConn(); msg != nil {
			out <- msg
			return
		}

		for {
			select {
			case <-ctx.Done():
				out <- newTCPMessageFailed(spec, startTime, ctx.Err())
				return

			case <-pollTicker.C:
				if msg := checkConn(); msg != nil {
					out <- msg
					return
				}
			}
		}
	}()

	return out
}

// OneTCP waits until a TCP connection can be made to an address, attempting a connection every
// defined interval. Both of these are contained in the given specifications. It also accepts a
// context function, which it uses to listen to cancellation events from the parent context.
// The returned channel is closed after the wait operation has finished or if the parent context is
// cancelled.
func OneTCP(spec *TCPSpec, waitTimeout time.Duration) <-chan *TCPMessage {
	return AllTCP([]*TCPSpec{spec}, waitTimeout)
}

// AllTCP waits until connections can be made to all given TCP input specifications for at most
// `waitTimeout` long. It returns a channel through which all wait operation-related messages will
// be sent.  The returned channel is closed after all wait operations have finished.
func AllTCP(specs []*TCPSpec, waitTimeout time.Duration) <-chan *TCPMessage {

	addrs := make([]string, len(specs))
	for i, spec := range specs {
		addrs[i] = spec.Addr()
	}

	var (
		chs         = make([](<-chan *TCPMessage), len(specs))
		out         = make(chan *TCPMessage)
		ctx, cancel = newContext()
	)

	for i, spec := range specs {
		chs[i] = singleTCP(ctx, spec)
	}

	msgs := merge(chs)
	timeout := time.NewTimer(waitTimeout)

	go func() {
		defer timeout.Stop()
		defer cancel()
		defer close(out)

		for {
			select {
			case <-timeout.C:
				msg := newTCPMessageFailed(
					nil,
					startTimeFromContext(ctx),
					fmt.Errorf("exceeded timeout limit of %s", waitTimeout),
				)
				out <- msg
				return

			case msg, isOpen := <-msgs:
				if !isOpen {
					return
				}
				out <- msg
			}
		}
	}()

	return out
}
